File: backend/app/pty/manager.py
from __future__ import annotations

import threading
import time
import uuid
from collections import deque
from queue import Queue
from typing import Any, Deque, Dict, List, Optional

from ..config import OUTPUT_QUEUE_MAXSIZE
from .reader import PTYOutputReader
from .session import PTYSession


class PTYManager:
    def __init__(self) -> None:
        self._lock = threading.RLock()
        self._session: Optional[PTYSession] = None
        self._session_id: Optional[str] = None
        self._started_at: Optional[float] = None

        self._output_queue: Queue[str] = Queue(maxsize=OUTPUT_QUEUE_MAXSIZE)
        self._history: Deque[str] = deque()
        self._history_bytes: List[int] = [0]

        self._reader: Optional[PTYOutputReader] = None

    @property
    def output_queue(self) -> Queue[str]:
        return self._output_queue

    def history_snapshot(self) -> List[str]:
        with self._lock:
            return list(self._history)

    def is_running(self) -> bool:
        with self._lock:
            return bool(self._session and self._session.is_alive())

    def start(
        self,
        command: str,
        args: Optional[List[str]] = None,
        cwd: Optional[str] = None,
        cols: int = 80,
        rows: int = 24,
    ) -> Dict[str, Any]:
        with self._lock:
            if self.is_running():
                raise RuntimeError("Session already running")

            self._history.clear()
            self._history_bytes[0] = 0
            while not self._output_queue.empty():
                try:
                    self._output_queue.get_nowait()
                except Exception:
                    break

            self._session = PTYSession(command=command, args=args, cwd=cwd, cols=cols, rows=rows)
            try:
                self._session.start()
            except Exception:
                self._session = None
                self._session_id = None
                self._started_at = None
                self._reader = None
                raise

            self._session_id = str(uuid.uuid4())
            self._started_at = time.time()

            self._reader = PTYOutputReader(
                session=self._session,
                output_queue=self._output_queue,
                history=self._history,
                history_bytes_ref=self._history_bytes,
            )
            self._reader.start()

            return {
                "session_id": self._session_id,
                "pid": self._session.pid,
                "cols": cols,
                "rows": rows,
            }

    def stop(self, force: bool = False) -> None:
        with self._lock:
            if not self._session:
                return

            if self._reader:
                try:
                    self._reader.stop()
                except Exception:
                    pass

            try:
                self._session.stop(force=force)
            finally:
                self._session = None
                self._session_id = None
                self._started_at = None

    def write(self, data: str) -> None:
        with self._lock:
            if not self._session or not self._session.is_alive():
                return
            self._session.write(data)

    def resize(self, cols: int, rows: int) -> None:
        with self._lock:
            if not self._session:
                return
            self._session.resize(cols, rows)

    def status(self) -> Dict[str, Any]:
        with self._lock:
            running = bool(self._session and self._session.is_alive())
            pid = self._session.pid if running and self._session else None
            dims = (
                {"cols": self._session.dimensions.cols, "rows": self._session.dimensions.rows}
                if self._session
                else {"cols": None, "rows": None}
            )
            uptime = int(time.time() - self._started_at) if running and self._started_at else 0

            return {
                "status": "running" if running else "stopped",
                "pid": pid,
                "uptime_seconds": uptime,
                "dimensions": dims,
                "session_id": self._session_id,
            }

    def shutdown(self) -> None:
        self.stop(force=True)


pty_manager = PTYManager()

File: backend-node/src/pty/manager.ts
import { v4 as uuidv4 } from 'uuid';
import { PTYSession } from './session.js';
import { stripTerminalIdentityResponses } from './sanitize.js';
import { HISTORY_MAX_BYTES, HISTORY_MAX_CHUNKS } from '../config.js';

interface PTYStartResult {
  session_id: string;
  pid: number | null;
  cols: number;
  rows: number;
}

interface PTYStatus {
  status: 'running' | 'stopped';
  pid: number | null;
  uptime_seconds: number;
  dimensions: { cols: number | null; rows: number | null };
  session_id: string | null;
  cwd: string | null;
}

type OutputListener = (chunk: string) => void;

class PTYManager {
  private session: PTYSession | null = null;
  private sessionId: string | null = null;
  private startedAt: number | null = null;
  private cwd: string | null = null;

  private outputListeners: OutputListener[] = [];

  private history: string[] = [];
  private historyBytes: number = 0;

  onOutput(listener: OutputListener): void {
    this.outputListeners.push(listener);
  }

  removeOutputListener(listener: OutputListener): void {
    const index = this.outputListeners.indexOf(listener);
    if (index > -1) {
      this.outputListeners.splice(index, 1);
    }
  }

  historySnapshot(): string[] {
    return [...this.history];
  }

  isRunning(): boolean {
    return this.session !== null && this.session.isAlive();
  }

  start(options: {
    command: string;
    args?: string[];
    cwd?: string;
    cols?: number;
    rows?: number;
  }): PTYStartResult {
    if (this.isRunning()) {
      throw new Error('Session already running');
    }

    this.history = [];
    this.historyBytes = 0;

    const cols = options.cols || 80;
    const rows = options.rows || 24;

    this.session = new PTYSession({
      command: options.command,
      args: options.args,
      cwd: options.cwd,
      cols,
      rows,
    });

    this.session.onData((chunk) => {
      const cleaned = stripTerminalIdentityResponses(chunk);
      if (!cleaned) return;

      this.pushHistory(cleaned);
      this.notifyOutputListeners(cleaned);
    });

    this.session.onExit(() => {
      this.notifyOutputListeners('\r\n[Process exited]\r\n');
    });

    this.session.start();

    this.sessionId = uuidv4();
    this.startedAt = Date.now();
    this.cwd = options.cwd || null;

    return {
      session_id: this.sessionId,
      pid: this.session.pid,
      cols,
      rows,
    };
  }

  private pushHistory(chunk: string): void {
    const bytes = Buffer.byteLength(chunk, 'utf-8');
    this.history.push(chunk);
    this.historyBytes += bytes;

    while (
      this.history.length > 0 &&
      (this.historyBytes > HISTORY_MAX_BYTES || this.history.length > HISTORY_MAX_CHUNKS)
    ) {
      const removed = this.history.shift()!;
      this.historyBytes -= Buffer.byteLength(removed, 'utf-8');
    }
  }

  private notifyOutputListeners(chunk: string): void {
    for (const listener of this.outputListeners) {
      try {
        listener(chunk);
      } catch {
        // Ignore listener errors
      }
    }
  }

  stop(force: boolean = false): void {
    if (!this.session) return;

    this.session.stop(force);
    this.session = null;
    this.sessionId = null;
    this.startedAt = null;
  }

  write(data: string): void {
    if (!this.session || !this.session.isAlive()) return;
    this.session.write(data);
  }

  resize(cols: number, rows: number): void {
    if (!this.session) return;
    this.session.resize(cols, rows);
  }

  status(): PTYStatus {
    const running = this.session !== null && this.session.isAlive();
    const pid = running && this.session ? this.session.pid : null;
    const dims = this.session
      ? { cols: this.session.dimensions.cols, rows: this.session.dimensions.rows }
      : { cols: null, rows: null };
    const uptime = running && this.startedAt ? Math.floor((Date.now() - this.startedAt) / 1000) : 0;

    return {
      status: running ? 'running' : 'stopped',
      pid,
      uptime_seconds: uptime,
      dimensions: dims,
      session_id: this.sessionId,
      cwd: this.cwd,
    };
  }

  shutdown(): void {
    this.stop(true);
  }
}

export const ptyManager = new PTYManager();
